File: backend-rails/lib/arc/graph_plan.rb
require "set"
module Arc
  class GraphPlan
    attr_reader :nodes, :incoming, :outgoing, :order, :ancestors, :available
    def initialize(definition)
      @nodes = definition.fetch("nodes").to_h { |n| [n.fetch("id"), n] }
      @incoming, @outgoing = Hash.new { |h, k| h[k] = [] }, Hash.new { |h, k| h[k] = [] }
      definition.fetch("edges").each { |e| @incoming[e["target"]] << e; @outgoing[e["source"]] << e }
      remaining = @nodes.to_h { |id, _| [id, @incoming[id].size] }
      ready = remaining.select { |_, count| count.zero? }.keys.sort
      @order, @ancestors, @available = [], {}, {}
      until ready.empty?
        id = ready.shift
        @order << @nodes[id]
        @ancestors[id] = @incoming[id].each_with_object(Set.new) { |e, set| set << e["source"]; set.merge(@ancestors.fetch(e["source"])) }
        @outgoing[id].each do |e|
          remaining[e["target"]] -= 1
          ready << e["target"] if remaining[e["target"]].zero?
        end
        ready.sort!
      end
      if @order.size != @nodes.size
        node = @nodes.values.find { |n| remaining[n["id"]] > 0 }
        raise ArcError.new(422, "Decision graphs cannot contain cycles").at_node(node)
      end
      availability(definition)
    end

    def availability(definition)
      logic, activation, predicates, scopes = Logic.new, {}, {}, {}
      @order.select { |n| n["type"] == "CONDITION" }.each_with_index { |n, i| predicates[n["id"]] = logic.make(i, 0, 1) }
      @order.each do |node|
        active = node["type"] == "INPUT" ? 1 : 0
        scope = node["type"] == "INPUT" ? definition["inputs"].to_h { |p| [p["name"], 1] } : {}
        @incoming[node["id"]].each do |edge|
          gate = activation.fetch(edge["source"])
          if edge["sourceHandle"] != "next" && predicates.key?(edge["source"])
            predicate = predicates[edge["source"]]
            gate = logic.apply(true, gate, edge["sourceHandle"] == "true" ? predicate : logic.invert(predicate))
          end
          active = logic.apply(false, active, gate)
          scopes.fetch(edge["source"]).each do |name, present|
            scope[name] = logic.apply(false, scope.fetch(name, 0), logic.apply(true, gate, present))
          end
        end
        @available[node["id"]] = active.zero? ? [] : scope.select { |_, present| logic.apply(true, active, logic.invert(present)).zero? }.keys.sort
        scope[node["output"]] = active if %w[FORMULA REFERENCE].include?(node["type"]) && node["output"]
        activation[node["id"]], scopes[node["id"]] = active, scope
      end
    end

    class Logic
      def initialize
        @branches, @unique, @cache, @inverse = [[2**31, 0, 0], [2**31, 1, 1]], {}, {}, { 0 => 1, 1 => 0 }
      end

      def make(variable, low, high)
        return low if low == high
        branch = [variable, low, high]
        return @unique[branch] if @unique.key?(branch)
        complex! if @branches.size >= 50_000
        @unique[branch] = @branches.size
        @branches << branch
        @unique[branch]
      end

      def invert(id)
        return @inverse[id] if @inverse.key?(id)
        variable, low, high = @branches[id]
        result = make(variable, invert(low), invert(high))
        @inverse[id], @inverse[result] = result, id
        result
      end

      def apply(conjunction, a, b)
        return a if a == b
        if conjunction
          return 0 if a.zero? || b.zero?
          return b if a == 1
          return a if b == 1
        else
          return 1 if a == 1 || b == 1
          return b if a.zero?
          return a if b.zero?
        end
        complex! if @cache.size > 200_000
        key = [conjunction, [a, b].min, [a, b].max]
        return @cache[key] if @cache.key?(key)
        x, y = @branches[a], @branches[b]
        top = [x[0], y[0]].min
        low = apply(conjunction, x[0] == top ? x[1] : a, y[0] == top ? y[1] : b)
        high = apply(conjunction, x[0] == top ? x[2] : a, y[0] == top ? y[2] : b)
        @cache[key] = make(top, low, high)
      end

      def complex!
        raise ArcError.new(422, "Branch analysis is too complex; split this graph into reusable rules")
      end
    end
  end
end

File: backend/src/main/java/dev/arc/engine/GraphPlan.java
package dev.arc.engine;

import dev.arc.error.ArcException;
import dev.arc.model.Definition;
import dev.arc.model.Definition.*;
import java.util.*;

/** A deterministic DAG order and branch-sensitive variable availability. */
public final class GraphPlan {
  public final Map<String, Node> nodes = new LinkedHashMap<>();
  public final Map<String, List<Edge>> incoming = new HashMap<>(), outgoing = new HashMap<>();
  public final List<Node> order = new ArrayList<>();
  public final Map<String, Set<String>> ancestors = new HashMap<>(),
      available = new LinkedHashMap<>();

  public GraphPlan(Definition d) {
    for (Node n : d.nodes()) nodes.put(n.id(), n);
    for (Edge e : d.edges()) {
      incoming.computeIfAbsent(e.target(), k -> new ArrayList<>()).add(e);
      outgoing.computeIfAbsent(e.source(), k -> new ArrayList<>()).add(e);
    }
    Map<String, Integer> remaining = new HashMap<>();
    Queue<String> ready = new PriorityQueue<>();
    for (Node n : d.nodes()) {
      int count = incoming.getOrDefault(n.id(), List.of()).size();
      remaining.put(n.id(), count);
      if (count == 0) ready.add(n.id());
    }
    while (!ready.isEmpty()) {
      String id = ready.remove();
      order.add(nodes.get(id));
      Set<String> predecessors = new HashSet<>();
      for (Edge e : incoming.getOrDefault(id, List.of())) {
        predecessors.add(e.source());
        predecessors.addAll(ancestors.get(e.source()));
      }
      ancestors.put(id, predecessors);
      for (Edge e : outgoing.getOrDefault(id, List.of()))
        if (remaining.merge(e.target(), -1, Integer::sum) == 0) ready.add(e.target());
    }
    if (order.size() != nodes.size()) {
      Node n =
          nodes.values().stream().filter(x -> remaining.get(x.id()) > 0).findFirst().orElseThrow();
      throw ArcException.invalid("Decision graphs cannot contain cycles")
          .atNode(null, null, n.id(), n.label());
    }
    availability(d);
  }

  private void availability(Definition d) {
    Logic logic = new Logic();
    Map<String, Integer> activation = new HashMap<>(), predicates = new HashMap<>();
    Map<String, Map<String, Integer>> scopes = new HashMap<>();
    int index = 0;
    for (Node n : order)
      if (n.type().equals("CONDITION")) predicates.put(n.id(), logic.variable(index++));
    for (Node n : order) {
      int active = n.type().equals("INPUT") ? 1 : 0;
      Map<String, Integer> scope = new HashMap<>();
      if (n.type().equals("INPUT")) for (Input p : d.inputs()) scope.put(p.name(), 1);
      for (Edge edge : incoming.getOrDefault(n.id(), List.of())) {
        int gate = activation.get(edge.source());
        if (!edge.sourceHandle().equals("next") && predicates.containsKey(edge.source())) {
          int test = predicates.get(edge.source());
          gate = logic.and(gate, edge.sourceHandle().equals("true") ? test : logic.not(test));
        }
        active = logic.or(active, gate);
        for (var entry : scopes.get(edge.source()).entrySet()) {
          int present = logic.and(gate, entry.getValue());
          scope.merge(entry.getKey(), present, logic::or);
        }
      }
      Set<String> guaranteed = new TreeSet<>();
      if (active != 0)
        for (var entry : scope.entrySet())
          if (logic.and(active, logic.not(entry.getValue())) == 0) guaranteed.add(entry.getKey());
      available.put(n.id(), guaranteed);
      if ((n.type().equals("FORMULA") || n.type().equals("REFERENCE")) && n.output() != null)
        scope.put(n.output(), active);
      activation.put(n.id(), active);
      scopes.put(n.id(), scope);
    }
  }

  // Reduced ordered Boolean decision diagrams avoid enumerating every combination
  // of conditions. Each condition is independent: validation never assumes its value.
  private static final class Logic {
    record Branch(int variable, int low, int high) {}

    record Operation(boolean conjunction, int a, int b) {}

    final List<Branch> branches =
        new ArrayList<>(
            List.of(new Branch(Integer.MAX_VALUE, 0, 0), new Branch(Integer.MAX_VALUE, 1, 1)));
    final Map<Branch, Integer> unique = new HashMap<>();
    final Map<Operation, Integer> cache = new HashMap<>();
    final Map<Integer, Integer> inverses = new HashMap<>(Map.of(0, 1, 1, 0));

    int make(int variable, int low, int high) {
      if (low == high) return low;
      Branch b = new Branch(variable, low, high);
      Integer found = unique.get(b);
      if (found != null) return found;
      if (branches.size() >= 50_000)
        throw ArcException.invalid(
            "Branch analysis is too complex; split this graph into reusable rules");
      int id = branches.size();
      branches.add(b);
      unique.put(b, id);
      return id;
    }

    int variable(int index) {
      return make(index, 0, 1);
    }

    int not(int id) {
      Integer result = inverses.get(id);
      if (result != null) return result;
      Branch b = branches.get(id);
      int inverse = make(b.variable(), not(b.low()), not(b.high()));
      inverses.put(id, inverse);
      inverses.put(inverse, id);
      return inverse;
    }

    int and(int a, int b) {
      return apply(true, a, b);
    }

    int or(int a, int b) {
      return apply(false, a, b);
    }

    int apply(boolean conjunction, int a, int b) {
      if (a == b) return a;
      if (conjunction) {
        if (a == 0 || b == 0) return 0;
        if (a == 1) return b;
        if (b == 1) return a;
      } else {
        if (a == 1 || b == 1) return 1;
        if (a == 0) return b;
        if (b == 0) return a;
      }
      if (cache.size() > 200_000)
        throw ArcException.invalid(
            "Branch analysis is too complex; split this graph into reusable rules");
      Operation key = new Operation(conjunction, Math.min(a, b), Math.max(a, b));
      Integer found = cache.get(key);
      if (found != null) return found;
      Branch x = branches.get(a), y = branches.get(b);
      int top = Math.min(x.variable(), y.variable());
      int low =
          apply(conjunction, x.variable() == top ? x.low() : a, y.variable() == top ? y.low() : b);
      int high =
          apply(
              conjunction, x.variable() == top ? x.high() : a, y.variable() == top ? y.high() : b);
      int result = make(top, low, high);
      cache.put(key, result);
      return result;
    }
  }
}
